File: lab04.1/ciphers_csharp/CaesarCipher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ciphers_csharp {
    internal static class CaesarCipher {
        private const string CYRILLIC_ALPHABET = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ";
        private const string LATIN_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

        private static string _caesarCipher(string oldMessage, int key, string alphabet) {
            string newMessage = "";
            foreach (char oldCh in oldMessage) {
                string currentAlphabet = char.IsUpper(oldCh) ? alphabet.ToUpper() : alphabet.ToLower();
                int oldChIndex = currentAlphabet.IndexOf(oldCh);
                char newCh;
                if (oldChIndex == -1) {
                    newCh = oldCh;
                }
                else {
                    int newChIndex = (oldChIndex + key) % currentAlphabet.Length;
                    if (newChIndex < 0) {
                        newChIndex = currentAlphabet.Length + newChIndex;
                    }
                    newCh = currentAlphabet[newChIndex];
                }
                newMessage += newCh;
            }
            return newMessage;
        }

        public static string CaesarCipherEncryption(string decryptedMessage, int key, string alphabet) {
            return _caesarCipher(decryptedMessage, key, alphabet);
        }

        public static string CaesarCipherDecryption(string encryptedMessage, int key, string alphabet) {
            return _caesarCipher(encryptedMessage, -key, alphabet);
        }

        public static string CaesarCipherCyrillicEncryption(string decryptedMessage, int key) {
            return CaesarCipherEncryption(decryptedMessage, key, CYRILLIC_ALPHABET);
        }

        public static string CaesarCipherCyrillicDecryption(string encryptedMessage, int key) {
            return CaesarCipherDecryption(encryptedMessage, key, CYRILLIC_ALPHABET);
        }

        public static string CaesarCipherLatinEncryption(string decryptedMessage, int key) {
            return CaesarCipherEncryption(decryptedMessage, key, LATIN_ALPHABET);
        }

        public static string CaesarCipherLatinDecryption(string encryptedMessage, int key) {
            return CaesarCipherDecryption(encryptedMessage, key, LATIN_ALPHABET);
        }
    }
}

File: lab04.1/ciphers_python/caesar_cipher.py
CYRILLIC_ALPHABET = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"
LATIN_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"


def _caesar_cipher(*, old_message: str, key: int, alphabet: str) -> str:
    new_message = ""
    for old_ch in old_message:
        current_alphabet = alphabet.upper() if old_ch.isupper() else alphabet.lower()
        old_ch_index = current_alphabet.find(old_ch)
        if (old_ch_index == -1):
            new_ch = old_ch
        else:
            new_ch_index = (old_ch_index + key) % len(current_alphabet)
            new_ch = current_alphabet[new_ch_index]
        new_message += new_ch
    return new_message


def caesar_cipher_encryption(*, decrypted_message: str, key: int, alphabet: str) -> str:
    return _caesar_cipher(
        old_message=decrypted_message, 
        key=key, 
        alphabet=alphabet,
    )


def caesar_cipher_decryption(*, encrypted_message: str, key: int, alphabet: str) -> str:
    return _caesar_cipher(
        old_message=encrypted_message, 
        key=-key, 
        alphabet=alphabet,
    )


def caesar_cipher_cyrillic_encryption(decrypted_message: str, key: int) -> str:
    return caesar_cipher_encryption(
        decrypted_message=decrypted_message, 
        key=key, 
        alphabet=CYRILLIC_ALPHABET,
    )


def caesar_cipher_cyrillic_decryption(encrypted_message: str, key: int) -> str:
    return caesar_cipher_decryption(
        encrypted_message=encrypted_message, 
        key=key, 
        alphabet=CYRILLIC_ALPHABET,
    )


def caesar_cipher_latin_encryption(decrypted_message: str, key: int) -> str:
    return caesar_cipher_encryption(
        decrypted_message=decrypted_message,
        key=key, 
        alphabet=LATIN_ALPHABET,
    )


def caesar_cipher_latin_decryption(encrypted_message: str, key: int) -> str:
    return caesar_cipher_decryption(
        encrypted_message=encrypted_message, 
        key=key, 
        alphabet=LATIN_ALPHABET,
    )
